### PR TITLE
OS-8231

### PR DESCRIPTION
**Pull request by danmcd**
Attempt to include piadm(1M)-generated `pi.rc` which displays available alternate platform-images to boot from on a bootable ZFS pool. See https://github.com/joyent/smartos-live/pull/993 for piadm(1M) changes.

### CODE REVIEW

**Comment by danmcd** [comment]
Sample generated `pi.rc` file with 7 total PIs.  Default + 6 means it's two pages worth (out of a high of 3). Note its use of the two routines under review here.

```
\
\ Generated by piadm(1M).
\

\ Assume mainmenu_options=4 for now.

set mainmenu_caption[5]="Alternate [P]latform Images..."
set mainmenu_command[5]="3 goto_menu"
set mainmenu_keycode[5]=112
set mainansi_caption[5]="Alternate ^[1mP^[mlatform Images..."

\ Will be reset by init_pi (see below and in menu-commands.4th).
set pimenu_optionstext="Platform Image: (UNINIT)"
\ For feeding init_pi.
set pitext="Platform Image: "

set menuset_name3="pi1"
set pi1menu_init[1]="init_pi"

set pi1menu_caption[1]="Back to Main Menu [Backspace]"
set pi1menu_command[1]="pi_draw_screen drop 1 goto_menu"
set pi1menu_keycode[1]=8
set pi1ansi_caption[1]="Back to Main Menu ^[1m[Backspace]^[m"

set pi1menu_caption[2]="[P]age: 1 of 2"
set pi1menu_command[2]="4 goto_menu"
set pi1menu_keycode[2]=112
set pi1ansi_caption[2]="^[1mP^[mage: 1 of 2"

set pi1menu_options=3
set pi1menu_optionstext="${pimenu_optionstext}"

set pi1menu_caption[3]="[D]efault (20210804T003855Z)"
set pi1menu_command[3]="s\" set bootpi=default\" evaluate pi_unload pi_draw_screen"
set pi1menu_keycode[3]=100
set pi1ansi_caption[3]="^[1mD^[mefault (20210804T003855Z)"


set pi1menu_caption[4]="20210729T002724Z"
set pi1menu_command[4]="pi_unload s\" set bootpi=20210729T002724Z\" evaluate s\" load /os/20210729T002724Z/platform/i86pc/kernel/amd64/unix\" evaluate s\" load -t rootfs /os/20210729T002724Z/platform/i86pc/amd64/boot_archive\" evaluate pi_draw_screen"
set pi1ansi_caption[4]="20210729T002724Z"


set pi1menu_caption[5]="20210715T010227Z"
set pi1menu_command[5]="pi_unload s\" set bootpi=20210715T010227Z\" evaluate s\" load /os/20210715T010227Z/platform/i86pc/kernel/amd64/unix\" evaluate s\" load -t rootfs /os/20210715T010227Z/platform/i86pc/amd64/boot_archive\" evaluate pi_draw_screen"
set pi1ansi_caption[5]="20210715T010227Z"


set pi1menu_caption[6]="20210701T204427Z"
set pi1menu_command[6]="pi_unload s\" set bootpi=20210701T204427Z\" evaluate s\" load /os/20210701T204427Z/platform/i86pc/kernel/amd64/unix\" evaluate s\" load -t rootfs /os/20210701T204427Z/platform/i86pc/amd64/boot_archive\" evaluate pi_draw_screen"
set pi1ansi_caption[6]="20210701T204427Z"


set pi1menu_caption[7]="20210617T001230Z"
set pi1menu_command[7]="pi_unload s\" set bootpi=20210617T001230Z\" evaluate s\" load /os/20210617T001230Z/platform/i86pc/kernel/amd64/unix\" evaluate s\" load -t rootfs /os/20210617T001230Z/platform/i86pc/amd64/boot_archive\" evaluate pi_draw_screen"
set pi1ansi_caption[7]="20210617T001230Z"


set pi1menu_caption[8]="20210520T001536Z"
set pi1menu_command[8]="pi_unload s\" set bootpi=20210520T001536Z\" evaluate s\" load /os/20210520T001536Z/platform/i86pc/kernel/amd64/unix\" evaluate s\" load -t rootfs /os/20210520T001536Z/platform/i86pc/amd64/boot_archive\" evaluate pi_draw_screen"
set pi1ansi_caption[8]="20210520T001536Z"

set menuset_name4="pi2"
set pi2menu_init[1]="init_pi"

set pi2menu_caption[1]="Back to Main Menu [Backspace]"
set pi2menu_command[1]="pi_draw_screen drop 1 goto_menu"
set pi2menu_keycode[1]=8
set pi2ansi_caption[1]="Back to Main Menu ^[1m[Backspace]^[m"

set pi2menu_caption[2]="[P]age: 2 of 2"
set pi2menu_command[2]="3 goto_menu"
set pi2menu_keycode[2]=112
set pi2ansi_caption[2]="^[1mP^[mage: 2 of 2"

set pi2menu_options=3
set pi2menu_optionstext="${pimenu_optionstext}"

set pi2menu_caption[3]="[D]efault (20210804T003855Z)"
set pi2menu_command[3]="s\" set bootpi=default\" evaluate pi_unload pi_draw_screen"
set pi2menu_keycode[3]=100
set pi2ansi_caption[3]="^[1mD^[mefault (20210804T003855Z)"


set pi2menu_caption[4]="20210422T002312Z"
set pi2menu_command[4]="pi_unload s\" set bootpi=20210422T002312Z\" evaluate s\" load /os/20210422T002312Z/platform/i86pc/kernel/amd64/unix\" evaluate s\" load -t rootfs /os/20210422T002312Z/platform/i86pc/amd64/boot_archive\" evaluate pi_draw_screen"
set pi2ansi_caption[4]="20210422T002312Z"

```

File: usr/src/boot/sys/boot/forth/menu-commands.4th [hunk 1]
```diff
@@ -138,6 +138,26 @@ create chaincmd 1030 chars allot
 	0 boot ( state -- )
 ;
 
+\
+\ Platform-image selection for standalone SmartOS is mostly in pi.rc.
+\ We steal a boot environment routine too in pi_draw_screen (way below).
+  
+
+: init_pi ( -- )
+	s" bootpi" getenv? if else
```
Comment (tsoome) [inline usr/src/boot/sys/boot/forth/menu-commands.4th:147]: getenv? 0= if  would let to drop 'else'.

Comment (tsoome) [reply usr/src/boot/sys/boot/forth/menu-commands.4th:147]: no space after getenv, that is, we have two words, getenv will return ( len addr | -1 ), getenv? will return ( -1 | 0), 0 is false, -1 is true. getenv? is defined in support.4th.

Comment (danmcd) [reply usr/src/boot/sys/boot/forth/menu-commands.4th:147]: Dammit I typed that, I know `getenv?` is a single word.  Uggh.  Will fix.

Comment (danmcd) [reply usr/src/boot/sys/boot/forth/menu-commands.4th:147]: Also there are other "if else" antipatterns in this introduced by illumos bugfix 12448.  A cleanup for later?

Comment (tsoome) [reply usr/src/boot/sys/boot/forth/menu-commands.4th]: hm, missed those on review. Yea, should clean those.

**Comment by danmcd** [comment]
Testing notes in ticket.

File: usr/src/boot/sys/boot/forth/menu-commands.4th [hunk 2]
```diff
@@ -138,6 +138,26 @@ create chaincmd 1030 chars allot
 	0 boot ( state -- )
 ;
 
+\
+\ Platform-image selection for standalone SmartOS is mostly in pi.rc.
+\ We steal a boot environment routine too in pi_draw_screen (way below).
```
Comment (bahamat) [inline usr/src/boot/sys/boot/forth/menu-commands.4th:143]: this comment feels a little clumsy. I'm not sure what it's saying. Maybe:

    \ We will also steal a boot environment routine later in pi_draw_screen (way below).

or

    \ We also steal a boot environment routine here that getsused in pi_draw_screen (way below).

Comment (danmcd) [reply usr/src/boot/sys/boot/forth/menu-commands.4th:143]: Using:
`\ We will also steal a boot environment routine in pi_draw_screen (way below).`

**Review by bahamat — APPROVED** [review]

**Pull request merged**